# S4 - Assessment Test 1

In [1]:
# Bibliotecas
import os
import sqlite3 as sql
import pandas as pd

## Funciones

In [2]:
# Función para tomar un archivo de consultas de SQL y formatear su texto
def query_text_format(path, file):
    # Cambio de directorio
    try:
        os.chdir(path)
    except FileNotFoundError:
        print(f"El directorio {path} no se encontró.")
        return
    
    # Lectura del archivo
    try:
        with open(file, 'r') as f:
            query_log = f.read()
    except FileNotFoundError:
        print(f"El archivo {file} no se encontró.")
        return
    
    # Formateo del texto
    query_log = query_log.replace(' */', '')
    query_log_lines = query_log.split('\n')    
    for i, line in enumerate(query_log_lines):
        query_log_lines[i] = line.strip()
        if line[:2] == '/*':
            query_log_lines[i] = line.replace('/*', '###')
        elif line != '':
            query_log_lines[i] = '\t' + line + ' \\'
        else:
            pass
        
    # Impresión del texto
    print('\n'.join(query_log_lines))

In [3]:
# Función para tomar el texto de una consulta SQL y ejecutarla
def execute_query(query_text, connection):
    # DataFrame desde consulta
    results_df = pd.read_sql_query(
        query_text,
        connection
    )
    # Impresión
    print(results_df)

## Configuración

In [4]:
# Limit removal for showing pandas.DataFrames' columns
pd.set_option('display.max_columns', None)
# Limit removal for showing pandas.DataFrames' rows
pd.set_option('display.max_rows', None)
# Modification of console with for displaying
#pd.set_option('display.width', 8000)

## Formateo de Consultas SQL

In [5]:
# Impresión de las consultas de SQL usando la función creada
path = 'G:\\15_Estudio\\Udemy\\SQL - Portilla Complete Bootcamp'
file = 'UDM-SQL-BTCMP--004.sql'

query_text_format(path, file)

### 30. GROUP BY pt. 2
### Explore the payment table
	SELECT * \
	FROM payment \
	LIMIT 5; \
### customer_id, staff_id, rental_id are integers, however
### they act as a categorical column.

### How much each customer spent.
	SELECT customer_id, SUM(amount) \
	FROM payment \
	GROUP BY customer_id \
	ORDER BY customer_id ASC; \

### Top 10 of customers who spent the most.
	SELECT customer_id, SUM(amount) \
	FROM payment \
	GROUP BY customer_id \
	ORDER BY SUM(amount) DESC \
	LIMIT 10; \

### Top 10 of customer with the most transactions
	SELECT customer_id, COUNT(amount) \
	FROM payment \
	GROUP BY customer_id \
	ORDER BY COUNT(amount) DESC \
	LIMIT 10; \

### Grouping by more than one variable; how many transactions?
	SELECT staff_id, customer_id, COUNT(amount) \
	FROM payment \
	GROUP BY staff_id, customer_id \
	ORDER BY staff_id ASC, customer_id ASC; \

### Grouping by a date column
### Need: conversion of timestamp to date
	SELECT DATE(payment_date) \
	FROM payment; \

### Dates wit

## Trabajo

### Conexión con la Base de Datos

In [6]:
# Conexión y cursor
conn = sql.connect('G:\\15_Estudio\\Udemy\\SQL - Portilla Complete Bootcamp\\dvdrental.db')
cur = conn.cursor()

### 30. GROUP BY pt. 2

In [7]:
# Explore the payment table
query = " \
	SELECT * \
	FROM payment \
	LIMIT 5; \
    "
execute_query(query, conn)
# customer_id, staff_id, rental_id are integers, however
# they act as a categorical column.

   payment_id  customer_id  staff_id  rental_id  amount  \
0       17503          341         2       1520    7.99   
1       17504          341         1       1778    1.99   
2       17505          341         1       1849    7.99   
3       17506          341         2       2829    2.99   
4       17507          341         2       3130    7.99   

                 payment_date  
0  2007-02-15 22:25:46.996577  
1  2007-02-16 17:23:14.996577  
2  2007-02-16 22:41:45.996577  
3  2007-02-19 19:39:56.996577  
4  2007-02-20 17:31:48.996577  


In [8]:
# How much each customer spent.
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	GROUP BY customer_id \
	ORDER BY customer_id ASC; \
    "
execute_query(query, conn)

     customer_id  SUM(amount)
0              1       114.70
1              2       123.74
2              3       130.76
3              4        81.78
4              5       134.65
5              6        84.75
6              7       130.72
7              8        85.77
8              9        78.80
9             10        94.76
10            11        99.77
11            12        93.74
12            13       131.73
13            14        96.77
14            15       134.68
15            16       109.75
16            17        87.82
17            18        78.81
18            19        98.82
19            20       103.73
20            21       146.68
21            22        94.81
22            23       106.73
23            24        85.77
24            25       104.73
25            26       146.68
26            27       123.70
27            28       105.70
28            29       138.65
29            30       123.66
30            31       104.74
31            32       112.74
32        

In [9]:
# Top 10 of customers who spent the most.
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	GROUP BY customer_id \
	ORDER BY SUM(amount) DESC \
	LIMIT 10; \
    "
execute_query(query, conn)

   customer_id  SUM(amount)
0          148       211.55
1          526       208.58
2          178       194.61
3          137       191.62
4          144       189.60
5          459       183.63
6          181       167.67
7          410       167.62
8          236       166.61
9          403       162.67


In [10]:
# Top 10 of customer with the most transactions
query = " \
	SELECT customer_id, COUNT(amount) \
	FROM payment \
	GROUP BY customer_id \
	ORDER BY COUNT(amount) DESC \
	LIMIT 10; \
    "
execute_query(query, conn)

   customer_id  COUNT(amount)
0          148             45
1          526             42
2          144             40
3          236             39
4          178             39
5           75             39
6          410             38
7          137             38
8          459             37
9          366             36


In [11]:
# Grouping by more than one variable; how many transactions?
query = " \
	SELECT staff_id, customer_id, COUNT(amount) \
	FROM payment \
	GROUP BY staff_id, customer_id \
	ORDER BY staff_id ASC, customer_id ASC; \
    "
execute_query(query, conn)

      staff_id  customer_id  COUNT(amount)
0            1            1             15
1            1            2             14
2            1            3             12
3            1            4             12
4            1            5             14
5            1            6             15
6            1            7             16
7            1            8             14
8            1            9             12
9            1           10             12
10           1           11             12
11           1           12              8
12           1           13             17
13           1           14             16
14           1           15             17
15           1           16             10
16           1           17              9
17           1           18              5
18           1           19             10
19           1           20             11
20           1           21             16
21           1           22             15
22         

In [12]:
# Grouping by a date column
# Need: conversion of timestamp to date
query = " \
	SELECT DATE(payment_date) \
	FROM payment; \
    "
execute_query(query, conn)

      DATE(payment_date)
0             2007-02-15
1             2007-02-16
2             2007-02-16
3             2007-02-19
4             2007-02-20
5             2007-02-21
6             2007-02-17
7             2007-02-20
8             2007-02-20
9             2007-02-16
10            2007-02-16
11            2007-02-17
12            2007-02-17
13            2007-02-18
14            2007-02-20
15            2007-02-21
16            2007-02-15
17            2007-02-15
18            2007-02-16
19            2007-02-15
20            2007-02-15
21            2007-02-16
22            2007-02-19
23            2007-02-17
24            2007-02-21
25            2007-02-21
26            2007-02-16
27            2007-02-18
28            2007-02-20
29            2007-02-20
30            2007-02-21
31            2007-02-16
32            2007-02-17
33            2007-02-18
34            2007-02-15
35            2007-02-15
36            2007-02-20
37            2007-02-20
38            2007-02-17


In [13]:
# Dates with most amount of money spent
query = " \
	SELECT DATE(payment_date), SUM(amount) \
	FROM payment \
	GROUP BY DATE(payment_date) \
	ORDER BY SUM(amount) DESC; \
    "
execute_query(query, conn)

   DATE(payment_date)  SUM(amount)
0          2007-04-30      5723.89
1          2007-03-21      2868.27
2          2007-03-01      2808.24
3          2007-04-29      2717.60
4          2007-03-18      2701.76
5          2007-04-27      2673.57
6          2007-03-20      2669.89
7          2007-04-28      2622.73
8          2007-03-19      2617.69
9          2007-03-22      2586.79
10         2007-03-02      2550.05
11         2007-03-17      2442.16
12         2007-03-23      2342.43
13         2007-04-08      2227.84
14         2007-04-06      2077.14
15         2007-04-09      2067.86
16         2007-04-07      1984.28
17         2007-04-10      1973.18
18         2007-04-11      1940.32
19         2007-04-12      1930.48
20         2007-02-19      1290.90
21         2007-02-18      1275.98
22         2007-02-20      1219.09
23         2007-02-15      1188.92
24         2007-02-17      1188.17
25         2007-02-16      1154.18
26         2007-02-21       917.87
27         2007-05-1

### 31. Challenge: GROUP BY

In [14]:
# 1. Staff member with the most payments handled
query = " \
	SELECT staff_id, COUNT(*) \
	FROM payment \
	GROUP BY staff_id \
	ORDER BY COUNT(*) DESC; \
    "
execute_query(query, conn)

   staff_id  COUNT(*)
0         2      7304
1         1      7292


In [15]:
# 2. Replacement cost per MPAA rating
query = " \
	SELECT rating, ROUND(AVG(replacement_cost), 2) \
	FROM film \
	GROUP BY rating \
	ORDER BY AVG(replacement_cost); \
    "
execute_query(query, conn)

  rating  ROUND(AVG(replacement_cost), 2)
0     PG                            18.96
1      G                            20.12
2  NC-17                            20.14
3      R                            20.23
4  PG-13                            20.40


In [16]:
# 3. Top 5 customer who spent the most
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	GROUP BY customer_id \
	ORDER BY SUM(amount) DESC \
	LIMIT 5; \
    "
execute_query(query, conn)

   customer_id  SUM(amount)
0          148       211.55
1          526       208.58
2          178       194.61
3          137       191.62
4          144       189.60


### 32. HAVING

In [17]:
# Explore the payment table
query = " \
	SELECT * \
	FROM payment \
	LIMIT 1; \
    "
execute_query(query, conn)

   payment_id  customer_id  staff_id  rental_id  amount  \
0       17503          341         2       1520    7.99   

                 payment_date  
0  2007-02-15 22:25:46.996577  


In [18]:
# Expenditure per customer, with filtered customers
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	WHERE customer_id NOT IN (184, 87, 477) \
	GROUP BY customer_id; \
    "
execute_query(query, conn)

     customer_id  SUM(amount)
0              1       114.70
1              2       123.74
2              3       130.76
3              4        81.78
4              5       134.65
5              6        84.75
6              7       130.72
7              8        85.77
8              9        78.80
9             10        94.76
10            11        99.77
11            12        93.74
12            13       131.73
13            14        96.77
14            15       134.68
15            16       109.75
16            17        87.82
17            18        78.81
18            19        98.82
19            20       103.73
20            21       146.68
21            22        94.81
22            23       106.73
23            24        85.77
24            25       104.73
25            26       146.68
26            27       123.70
27            28       105.70
28            29       138.65
29            30       123.66
30            31       104.74
31            32       112.74
32        

In [19]:
### Adding filtering on expenditure
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	WHERE customer_id NOT IN (184, 87, 477) \
	GROUP BY customer_id \
	HAVING SUM(amount) > 100 \
	ORDER BY SUM(amount) DESC; \
    "
execute_query(query, conn)

     customer_id  SUM(amount)
0            148       211.55
1            526       208.58
2            178       194.61
3            137       191.62
4            144       189.60
5            459       183.63
6            181       167.67
7            410       167.62
8            236       166.61
9            403       162.67
10           522       161.68
11           469       158.65
12           470       157.69
13           373       156.66
14           259       154.70
15           468       154.66
16           462       152.69
17           372       152.68
18           187       151.73
19           550       151.69
20           176       151.68
21           532       149.69
22            75       149.61
23           209       147.71
24            26       146.68
25            21       146.68
26            86       146.68
27           211       146.67
28           346       145.70
29           366       145.64
30            50       144.70
31           360       144.68
32        

In [20]:
# Explore customer table
query = " \
	SELECT * \
	FROM customer \
	LIMIT 1; \
    "
execute_query(query, conn)

   customer_id  store_id first_name last_name                         email  \
0          524         1      Jared       Ely  jared.ely@sakilacustomer.org   

   address_id activebool create_date              last_update  active  
0         530          t  2006-02-14  2013-05-26 14:49:45.738       1  


In [21]:
# Number of customer associated to a store
query = " \
	SELECT store_id, COUNT(*) \
	FROM customer \
	GROUP BY store_id \
	ORDER BY COUNT(*) DESC; \
    "
execute_query(query, conn)

   store_id  COUNT(*)
0         1       326
1         2       273


In [22]:
# Filtering results to stores with a certain lower number of customers
query = " \
	SELECT store_id, COUNT(customer_id) \
	FROM customer \
	GROUP BY store_id \
	HAVING COUNT(customer_id) > 300 \
	ORDER BY COUNT(customer_id) DESC; \
    "
execute_query(query, conn)

   store_id  COUNT(customer_id)
0         1                 326


### 33. Challenge: HAVING

In [23]:
# 1. Find the customers with 40 or more transactions
query = " \
	SELECT * \
	FROM payment \
	LIMIT 1; \
    "
execute_query(query, conn)

   payment_id  customer_id  staff_id  rental_id  amount  \
0       17503          341         2       1520    7.99   

                 payment_date  
0  2007-02-15 22:25:46.996577  


In [24]:
query = " \
    SELECT customer_id, COUNT(payment_id) \
	FROM payment \
	GROUP BY customer_id \
	HAVING COUNT(payment_id) >= 40 \
	ORDER BY COUNT(payment_id) DESC; \
    "
execute_query(query, conn)

   customer_id  COUNT(payment_id)
0          148                 45
1          526                 42
2          144                 40


In [25]:
## 2. Which customers spent more than $100 in payments
## associated to staff member #2
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	WHERE staff_id = 2 \
	GROUP BY customer_id \
	HAVING SUM(amount) > 100 \
	ORDER BY SUM(amount) DESC; \
    "
execute_query(query, conn)

   customer_id  SUM(amount)
0          187       110.81
1          148       110.78
2          211       108.77
3          522       102.80
4          526       101.78


### 34, 35, 36. Assessment Test 1

#### Complete the following tasks:

	1. Return the customer IDs of customers who have spent at least $110 with the staff member who has an ID of 2.
	The answer should be customers 187 and 148.

	2. How many films begin with the letter J?
	The answer should be 20.

	3. What customer has the highest customer ID number whose name starts with an 'E' and has an address ID lower than 500? \
	The answer is Eddie Tomlin.

#### Queries

In [26]:
# 1. Return the customer IDs of customers who have spent
# at least $110 with the staff member who has an ID of 2.
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	WHERE staff_id = 2 \
	GROUP BY customer_id \
	HAVING SUM(amount) >= 110 \
	ORDER BY SUM(amount) DESC; \
    "
execute_query(query, conn)

   customer_id  SUM(amount)
0          187       110.81
1          148       110.78


In [27]:
# 2. How many films begin with the letter J?
query = " \
	SELECT * \
	FROM film \
	LIMIT 1; \
    "
execute_query(query, conn)

   film_id            title  \
0      133  Chamber Italian   

                                         description  release_year  \
0  A Fateful Reflection of a Moose And a Husband ...          2006   

   language_id  rental_duration  rental_rate  length  replacement_cost rating  \
0            1                7         4.99     117             14.99  NC-17   

               last_update special_features  \
0  2013-05-26 14:50:58.951       {Trailers}   

                                            fulltext  
0  'chamber':1 'fate':4 'husband':11 'italian':2 ...  


In [28]:
query = " \
	SELECT COUNT(film_id) \
	FROM film \
	WHERE title LIKE 'J%'; \
    "
# LIKE 'J%' actually was ILIKE 'j%'
execute_query(query, conn)

   COUNT(film_id)
0              20


In [29]:
# 3. What customer has the highest customer ID number whose
# name starts with an 'E' and has an address ID lower than 500?
query = " \
	SELECT * \
	FROM customer; \
    "
execute_query(query, conn)

     customer_id  store_id   first_name     last_name  \
0            524         1        Jared           Ely   
1              1         1         Mary         Smith   
2              2         1     Patricia       Johnson   
3              3         1        Linda      Williams   
4              4         2      Barbara         Jones   
5              5         1    Elizabeth         Brown   
6              6         2     Jennifer         Davis   
7              7         1        Maria        Miller   
8              8         2        Susan        Wilson   
9              9         2     Margaret         Moore   
10            10         1      Dorothy        Taylor   
11            11         2         Lisa      Anderson   
12            12         1        Nancy        Thomas   
13            13         2        Karen       Jackson   
14            14         2        Betty         White   
15            15         1        Helen        Harris   
16            16         2     

598  2006-02-14  2013-05-26 14:49:45.738       1  


In [30]:
query = " \
	SELECT first_name, last_name \
	FROM customer \
	WHERE first_name LIKE 'E%' \
		AND address_id < 500 \
	ORDER BY customer_id DESC \
	LIMIT 1; \
    "
# LIKE 'E%' actually was ILIKE 'e%'
execute_query(query, conn)

  first_name last_name
0      Eddie    Tomlin


### 37. Overview of JOINS
Blank/in notes

### 38. Introduction to JOINs
Blank/in notes

### 39. AS Statement

In [31]:
query = " \
	SELECT amount AS rental_price \
	FROM payment; \
    "
execute_query(query, conn)

       rental_price
0              7.99
1              1.99
2              7.99
3              2.99
4              7.99
5              5.99
6              5.99
7              5.99
8              2.99
9              4.99
10             6.99
11             0.99
12             0.99
13             6.99
14             8.99
15             0.99
16             3.99
17             4.99
18             0.99
19             0.99
20             4.99
21             0.99
22             4.99
23             5.99
24             2.99
25             2.99
26             8.99
27             0.99
28             4.99
29             8.99
30             7.99
31             2.99
32             8.99
33             0.99
34             2.99
35             0.99
36             6.99
37             8.99
38             3.99
39             0.99
40             2.99
41             0.99
42             5.99
43             0.99
44             2.99
45             0.99
46             4.99
47             4.99
48             4.99


In [32]:
query = " \
	SELECT SUM(amount) AS net_revenue \
	FROM payment; \
    "
execute_query(query, conn)

   net_revenue
0     61312.04


In [33]:
# Aliases do not allow usage for filtering
query = " \
	SELECT COUNT(amount) \
	FROM payment; \
    "
execute_query(query, conn)

   COUNT(amount)
0          14596


In [34]:
query = " \
	SELECT COUNT(amount) AS num_transactions \
	FROM payment; \
    "
# An alias was added
execute_query(query, conn)

   num_transactions
0             14596


In [35]:
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	GROUP BY customer_id; \
    "
# Aggregation added, no alias
# Grouping added
execute_query(query, conn)

     customer_id  SUM(amount)
0              1       114.70
1              2       123.74
2              3       130.76
3              4        81.78
4              5       134.65
5              6        84.75
6              7       130.72
7              8        85.77
8              9        78.80
9             10        94.76
10            11        99.77
11            12        93.74
12            13       131.73
13            14        96.77
14            15       134.68
15            16       109.75
16            17        87.82
17            18        78.81
18            19        98.82
19            20       103.73
20            21       146.68
21            22        94.81
22            23       106.73
23            24        85.77
24            25       104.73
25            26       146.68
26            27       123.70
27            28       105.70
28            29       138.65
29            30       123.66
30            31       104.74
31            32       112.74
32        

In [36]:
query = " \
	SELECT customer_id, SUM(amount) AS total_spent \
	FROM payment \
	GROUP BY customer_id; \
    "
# Alias added 
execute_query(query, conn)

     customer_id  total_spent
0              1       114.70
1              2       123.74
2              3       130.76
3              4        81.78
4              5       134.65
5              6        84.75
6              7       130.72
7              8        85.77
8              9        78.80
9             10        94.76
10            11        99.77
11            12        93.74
12            13       131.73
13            14        96.77
14            15       134.68
15            16       109.75
16            17        87.82
17            18        78.81
18            19        98.82
19            20       103.73
20            21       146.68
21            22        94.81
22            23       106.73
23            24        85.77
24            25       104.73
25            26       146.68
26            27       123.70
27            28       105.70
28            29       138.65
29            30       123.66
30            31       104.74
31            32       112.74
32        

In [37]:
# Alias removed 
# Added HAVING 
query = " \
	SELECT customer_id, SUM(amount) \
	FROM payment \
	GROUP BY customer_id \
	HAVING SUM(amount) > 100; \
    "
execute_query(query, conn)

     customer_id  SUM(amount)
0              1       114.70
1              2       123.74
2              3       130.76
3              5       134.65
4              7       130.72
5             13       131.73
6             15       134.68
7             16       109.75
8             20       103.73
9             21       146.68
10            23       106.73
11            25       104.73
12            26       146.68
13            27       123.70
14            28       105.70
15            29       138.65
16            30       123.66
17            31       104.74
18            32       112.74
19            35       118.70
20            38       127.66
21            39       141.71
22            40       105.74
23            41       118.75
24            42       111.71
25            44       101.79
26            45       124.74
27            46       131.69
28            49       116.76
29            50       144.70
30            51       123.70
31            52       106.72
32        

295          595       110.71


In [38]:
# Alias added 
query = " \
	SELECT customer_id, SUM(amount) AS total_spent \
	FROM payment \
	GROUP BY customer_id \
	HAVING SUM(amount) > 100; \
    "
execute_query(query, conn)

     customer_id  total_spent
0              1       114.70
1              2       123.74
2              3       130.76
3              5       134.65
4              7       130.72
5             13       131.73
6             15       134.68
7             16       109.75
8             20       103.73
9             21       146.68
10            23       106.73
11            25       104.73
12            26       146.68
13            27       123.70
14            28       105.70
15            29       138.65
16            30       123.66
17            31       104.74
18            32       112.74
19            35       118.70
20            38       127.66
21            39       141.71
22            40       105.74
23            41       118.75
24            42       111.71
25            44       101.79
26            45       124.74
27            46       131.69
28            49       116.76
29            50       144.70
30            51       123.70
31            52       106.72
32        

In [39]:
# The following should return error
# Erroneous filtering by alias
query = " \
	SELECT customer_id, SUM(amount) AS total_spent \
	FROM payment \
	GROUP BY customer_id \
	HAVING total_spent > 100; \
    "
execute_query(query, conn)
# Apparently, that's not the case in SQLite
# SQLite seems to support using aliases defined in the same query inside HAVING

     customer_id  total_spent
0              1       114.70
1              2       123.74
2              3       130.76
3              5       134.65
4              7       130.72
5             13       131.73
6             15       134.68
7             16       109.75
8             20       103.73
9             21       146.68
10            23       106.73
11            25       104.73
12            26       146.68
13            27       123.70
14            28       105.70
15            29       138.65
16            30       123.66
17            31       104.74
18            32       112.74
19            35       118.70
20            38       127.66
21            39       141.71
22            40       105.74
23            41       118.75
24            42       111.71
25            44       101.79
26            45       124.74
27            46       131.69
28            49       116.76
29            50       144.70
30            51       123.70
31            52       106.72
32        

In [40]:
# Another (supposed) erroneous query
query = " \
	SELECT customer_id, amount AS new_name \
	FROM payment \
	WHERE new_name > 2; \
    "
execute_query(query, conn)
# Case similar to the previous one

       customer_id  new_name
0              341      7.99
1              341      7.99
2              341      2.99
3              341      7.99
4              341      5.99
5              342      5.99
6              342      5.99
7              342      2.99
8              343      4.99
9              343      6.99
10             343      6.99
11             343      8.99
12             344      3.99
13             344      4.99
14             345      4.99
15             345      4.99
16             346      5.99
17             346      2.99
18             346      2.99
19             347      8.99
20             347      4.99
21             347      8.99
22             347      7.99
23             348      2.99
24             348      8.99
25             349      2.99
26             349      6.99
27             349      8.99
28             350      3.99
29             350      2.99
30             351      5.99
31             351      2.99
32             352      4.99
33            

### 40. INNER JOIN

In [41]:
query = " \
	SELECT * \
	FROM payment \
	LIMIT 1; \
    "
execute_query(query, conn)

   payment_id  customer_id  staff_id  rental_id  amount  \
0       17503          341         2       1520    7.99   

                 payment_date  
0  2007-02-15 22:25:46.996577  


In [42]:
query = " \
	SELECT * \
	FROM customer \
	LIMIT 1; \
    "
execute_query(query, conn)

   customer_id  store_id first_name last_name                         email  \
0          524         1      Jared       Ely  jared.ely@sakilacustomer.org   

   address_id activebool create_date              last_update  active  
0         530          t  2006-02-14  2013-05-26 14:49:45.738       1  


In [43]:
query = " \
	SELECT * \
	FROM payment \
	INNER JOIN customer \
		ON payment.customer_id = customer.customer_id \
    LIMIT 20; \
    "
execute_query(query, conn)

    payment_id  customer_id  staff_id  rental_id  amount  \
0        17503          341         2       1520    7.99   
1        17504          341         1       1778    1.99   
2        17505          341         1       1849    7.99   
3        17506          341         2       2829    2.99   
4        17507          341         2       3130    7.99   
5        17508          341         1       3382    5.99   
6        17509          342         2       2190    5.99   
7        17510          342         1       2914    5.99   
8        17511          342         1       3081    2.99   
9        17512          343         2       1547    4.99   
10       17513          343         1       1564    6.99   
11       17514          343         2       1879    0.99   
12       17515          343         2       1922    0.99   
13       17516          343         2       2461    6.99   
14       17517          343         1       2980    8.99   
15       17518          343         1   

In [44]:
# Querying for specific columns 
query = " \
	SELECT payment_id, payment.customer_id, first_name \
	FROM payment \
	INNER JOIN customer \
		ON payment.customer_id = customer.customer_id \
    LIMIT 20; \
    "
execute_query(query, conn)

    payment_id  customer_id first_name
0        17503          341      Peter
1        17504          341      Peter
2        17505          341      Peter
3        17506          341      Peter
4        17507          341      Peter
5        17508          341      Peter
6        17509          342     Harold
7        17510          342     Harold
8        17511          342     Harold
9        17512          343    Douglas
10       17513          343    Douglas
11       17514          343    Douglas
12       17515          343    Douglas
13       17516          343    Douglas
14       17517          343    Douglas
15       17518          343    Douglas
16       17519          344      Henry
17       17520          344      Henry
18       17521          344      Henry
19       17522          345       Carl


### 41. FULL OUTER JOIN

In [45]:
query = " \
	SELECT * \
	FROM payment \
	FULL OUTER JOIN customer \
		ON payment.customer_id = customer.customer_id \
	WHERE customer.customer_id IS null \
		OR payment.payment_id IS null \
    LIMIT 20; \
    "
execute_query(query, conn)

Empty DataFrame
Columns: [payment_id, customer_id, staff_id, rental_id, amount, payment_date, customer_id, store_id, first_name, last_name, email, address_id, activebool, create_date, last_update, active]
Index: []
